File: commands/blackjack.js
const Discord = require('discord.js');
const { Command } = require('./classes/command');
const { format } = require('../utils/math');
const { checkValidBet } = require('../utils/betting');

class Blackjack extends Command {
  constructor(client) {
    super(client, 'blackjack', 'bj with silverwolf', [
      {
        name: 'amount',
        description: 'the amount of mystic credits to bet',
        type: 3,
        required: true,
      },
    ]);
  }

  async run(interaction) {
    const amountString = interaction.options.getString('amount');
    const amount = await checkValidBet(interaction, amountString);
    if (amount === null) {
      return;
    }

    const deck = this.createDeck();
    const playerHand = [this.drawCard(deck), this.drawCard(deck)];
    const dealerHand = [this.drawCard(deck), this.drawCard(deck)];

    const gameMessage = await interaction.editReply({
      embeds: [this.buildEmbed(playerHand, dealerHand, 'Game Start')],
      components: [this.buildButtons()],
    });

    const collector = gameMessage.createMessageComponentCollector({ time: 60000 });

    collector.on('collect', async (i) => {
      if (i.user.id !== interaction.user.id) {
        await i.reply({ content: "These buttons aren't for you smh", ephemeral: true });
        return;
      }

      if (i.customId === 'hit') {
        playerHand.push(this.drawCard(deck));
        const playerTotal = this.calculateHand(playerHand);
        if (playerTotal > 21) {
          collector.stop('busted');
        } else {
          await i.update({ embeds: [this.buildEmbed(playerHand, dealerHand, 'Hit')], components: [this.buildButtons()] });
        }
      } else if (i.customId === 'stand') {
        collector.stop('stand');
      }
    });

    collector.on('end', async (collected, reason) => {
      if (reason === 'time') {
        await interaction.editReply({ content: 'Time ran out! The game has ended.', components: [] });
        await this.handleLoss(interaction, amount, playerHand, dealerHand, 'You ran out of time! (No, soren, you cant do that anymore)');
        return;
      }

      let resultMessage;
      if (reason === 'busted') {
        await this.handleLoss(interaction, amount, playerHand, dealerHand, 'You busted!');
        return;
      }

      while (this.calculateHand(dealerHand) < 17) {
        dealerHand.push(this.drawCard(deck));
      }

      const playerTotal = this.calculateHand(playerHand);
      const dealerTotal = this.calculateHand(dealerHand);

      if (dealerTotal > 21 || playerTotal > dealerTotal) {
        resultMessage = 'You win!';
        await this.handleWin(interaction, amount, playerHand, dealerHand, resultMessage);
      } else if (playerTotal < dealerTotal) {
        resultMessage = 'Silverwolf wins!';
        await this.handleLoss(interaction, amount, playerHand, dealerHand, resultMessage);
      } else {
        resultMessage = 'No one wins!';
        await this.handleTie(interaction, amount, playerHand, dealerHand, resultMessage);
      }
    });
  }

  createDeck() {
    const suits = ['♠', '♣', '♥', '♦'];
    const values = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'];
    const deck = [];

    suits.forEach((suit) => {
      values.forEach((value) => {
        deck.push({ suit, value });
      });
    });

    // Shuffle the deck
    for (let i = deck.length - 1; i > 0; i -= 1) {
      const j = Math.floor(Math.random() * (i + 1));
      [deck[i], deck[j]] = [deck[j], deck[i]];
    }

    return deck;
  }

  drawCard(deck) {
    return deck.pop();
  }

  calculateHand(hand) {
    let total = 0;
    let aces = 0;

    hand.forEach((card) => {
      if (card.value === 'A') {
        aces += 1;
        total += 11;
      } else if (['K', 'Q', 'J'].includes(card.value)) {
        total += 10;
      } else {
        total += parseInt(card.value, 10);
      }
    });

    while (total > 21 && aces > 0) {
      total -= 10;
      aces -= 1;
    }

    return total;
  }

  buildEmbed(playerHand, dealerHand, title) {
    return new Discord.EmbedBuilder()
      .setColor('#0099ff')
      .setTitle(`Blackjack - ${title}`)
      .setDescription(`Your hand: ${this.formatHand(playerHand)} (${this.calculateHand(playerHand)})\nSilverwolf's hand: ${this.formatHand([dealerHand[0]])} (??)\n\nHit or Stand?`);
  }

  buildButtons() {
    return new Discord.ActionRowBuilder()
      .addComponents(
        new Discord.ButtonBuilder()
          .setCustomId('hit')
          .setLabel('Hit')
          .setStyle(Discord.ButtonStyle.Primary),
        new Discord.ButtonBuilder()
          .setCustomId('stand')
          .setLabel('Stand')
          .setStyle(Discord.ButtonStyle.Secondary),
      );
  }

  formatHand(hand) {
    return hand.map((card) => `${card.suit}${card.value}`).join(', ');
  }

  async handleWin(interaction, amount, playerHand, dealerHand, message) {
    let streak = await this.client.db.user.getUserAttr(interaction.user.id, 'blackjackStreak');
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackTimesPlayed', 1);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackAmountGambled', amount);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackTimesWon', 1);

    const multiplier = await this.client.db.marriage.getMarriageBenefits(interaction.user.id) * 2.1 * 1.08 ** streak;
    streak += 1;
    if (streak > await this.client.db.user.getUserAttr(interaction.user.id, 'blackjackMaxStreak')) {
      await this.client.db.user.setUserAttr(interaction.user.id, 'blackjackMaxStreak', streak);
    }

    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackAmountWon', amount * multiplier);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackRelativeWon', multiplier);
    await this.client.db.user.addUserAttr(interaction.user.id, 'credits', amount * multiplier - amount);
    await this.client.db.user.setUserAttr(interaction.user.id, 'blackjackStreak', streak);

    await interaction.editReply({
      embeds: [new Discord.EmbedBuilder()
        .setColor('#00AA00')
        .setTitle(`${message} You won ${format(amount * multiplier)} mystic credits!`)
        .setDescription(`Your hand: ${this.formatHand(playerHand)} (${this.calculateHand(playerHand)})
Silverwolf's hand: ${this.formatHand(dealerHand)} (${this.calculateHand(dealerHand)})
You are now on a streak of ${streak}`)],
      components: [],
    });
  }

  async handleLoss(interaction, amount, playerHand, dealerHand, message) {
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackTimesPlayed', 1);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackAmountGambled', amount);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackTimesLost', 1);
    await this.client.db.user.addUserAttr(interaction.user.id, 'credits', -amount);
    await this.client.db.user.setUserAttr(interaction.user.id, 'blackjackStreak', 0);

    await interaction.editReply({
      embeds: [new Discord.EmbedBuilder()
        .setColor('#AA0000')
        .setTitle(`${message} You lost ${format(amount)} mystic credits!`)
        .setDescription(`Your hand: ${this.formatHand(playerHand)} (${this.calculateHand(playerHand)})
Silverwolf's hand: ${this.formatHand(dealerHand)} (${this.calculateHand(dealerHand)})`)],
      components: [],
    });
  }

  async handleTie(interaction, amount, playerHand, dealerHand, message) {
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackTimesPlayed', 1);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackAmountGambled', amount);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackTimesDrawn', 1);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackAmountWon', amount);
    await this.client.db.user.addUserAttr(interaction.user.id, 'blackjackRelativeWon', 1);
    await interaction.editReply({
      embeds: [new Discord.EmbedBuilder()
        .setColor('#FFFF00')
        .setTitle(`${message} Nothing happened to your ${format(amount)} mystic credits, boring.`)
        .setDescription(`Your hand: ${this.formatHand(playerHand)} (${this.calculateHand(playerHand)})
Silverwolf's hand: ${this.formatHand(dealerHand)} (${this.calculateHand(dealerHand)})`)],
      components: [],
    });
  }
}

module.exports = Blackjack;

File: src/commands/blackjack.rs
use crate::{Context, Error, utils};
use poise::serenity_prelude as serenity;
use rand::seq::SliceRandom;
use futures::StreamExt;

#[derive(Clone, Copy)]
pub struct Card {
    pub suit: &'static str,
    pub value: &'static str,
}

impl Card {
    fn score(&self) -> i64 {
        match self.value {
            "A" => 11,
            "K" | "Q" | "J" | "10" => 10,
            _ => self.value.parse().unwrap_or(0),
        }
    }
}

fn create_deck() -> Vec<Card> {
    let suits = ["♠", "♣", "♥", "♦"];
    let values = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"];
    let mut deck = Vec::new();
    for suit in suits {
        for value in values {
            deck.push(Card { suit, value });
        }
    }
    let mut rng = rand::thread_rng();
    deck.shuffle(&mut rng);
    deck
}

fn calculate_hand(hand: &[Card]) -> i64 {
    let mut total = hand.iter().map(|c| c.score()).sum::<i64>();
    let mut aces = hand.iter().filter(|c| c.value == "A").count();
    while total > 21 && aces > 0 {
        total -= 10;
        aces -= 1;
    }
    total
}

fn format_hand(hand: &[Card]) -> String {
    hand.iter().map(|c| format!("{}{}", c.suit, c.value)).collect::<Vec<_>>().join(", ")
}

/// bj with silverwolf
#[poise::command(slash_command)]
pub async fn blackjack(
    ctx: Context<'_>,
    #[description = "the amount of mystic credits to bet"] amount_str: String,
) -> Result<(), Error> {
    let bet_res = utils::check_valid_bet_raw(ctx.data().db.as_ref(), &ctx.author().id.to_string(), &amount_str).await?;
    let amount = match bet_res {
        utils::BetResult::Valid(a) => a,
        utils::BetResult::Infinity => {
            ctx.send(poise::CreateReply::default()
                .embed(serenity::CreateEmbed::new()
                    .color(0xAA0000)
                    .title("You have been spotted cheating! Mystic credits set to 0."))
            ).await?;
            return Ok(());
        }
        _ => {
            ctx.say("Invalid bet amount!").await?;
            return Ok(());
        }
    };

    let mut deck = create_deck();
    let mut player_hand = vec![deck.pop().unwrap(), deck.pop().unwrap()];
    let mut dealer_hand = vec![deck.pop().unwrap(), deck.pop().unwrap()];

    let ctx_id = ctx.id();
    let hit_id = format!("{}hit", ctx_id);
    let stand_id = format!("{}stand", ctx_id);

    let build_embed = |p_hand: &[Card], d_hand: &[Card], title: &str, finished: bool| {
        let d_str = if finished { format_hand(d_hand) } else { format_hand(&d_hand[0..1]) };
        let d_score = if finished { calculate_hand(d_hand).to_string() } else { "??".to_string() };

        serenity::CreateEmbed::new()
            .color(0x0099ff)
            .title(format!("Blackjack - {}", title))
            .description(format!("Your hand: {} ({})\nSilverwolf's hand: {} ({})\n\n{}", 
                format_hand(p_hand), 
                calculate_hand(p_hand),
                d_str,
                d_score,
                if finished { "" } else { "Hit or Stand?" }
            ))
    };

    let reply = ctx.send(poise::CreateReply::default()
        .embed(build_embed(&player_hand, &dealer_hand, "Game Start", false))
        .components(vec![serenity::CreateActionRow::Buttons(vec![
            serenity::CreateButton::new(&hit_id).label("Hit").style(serenity::ButtonStyle::Primary),
            serenity::CreateButton::new(&stand_id).label("Stand").style(serenity::ButtonStyle::Secondary),
        ])])
    ).await?;

    let mut interaction_stream = reply
        .message()
        .await?
        .await_component_interactions(ctx.serenity_context())
        .timeout(std::time::Duration::from_secs(60))
        .stream();

    let mut finished = false;
    let mut reason = "time";

    while let Some(mci) = interaction_stream.next().await {
        if mci.user.id != ctx.author().id {
            mci.create_response(ctx.serenity_context(), serenity::CreateInteractionResponse::Message(
                serenity::CreateInteractionResponseMessage::new().content("These buttons aren't for you smh").ephemeral(true)
            )).await?;
            continue;
        }

        if mci.data.custom_id == hit_id {
            player_hand.push(deck.pop().unwrap());
            if calculate_hand(&player_hand) > 21 {
                reason = "busted";
                finished = true;
                break;
            } else {
                mci.create_response(ctx.serenity_context(), serenity::CreateInteractionResponse::UpdateMessage(
                    serenity::CreateInteractionResponseMessage::new()
                        .embed(build_embed(&player_hand, &dealer_hand, "Hit", false))
                )).await?;
            }
        } else if mci.data.custom_id == stand_id {
            reason = "stand";
            finished = true;
            break;
        }
    }

    if !finished {
        ctx.say("Time ran out!").await?;
        // Handle loss...
        return Ok(());
    }

    if reason == "busted" {
        ctx.data().db.record_blackjack_loss(&ctx.author().id.to_string(), amount).await?;
        ctx.send(poise::CreateReply::default()
            .embed(serenity::CreateEmbed::new()
                .color(0xAA0000)
                .title(format!("You busted! You lost {} mystic credits!", utils::format(amount)))
                .description(format!("Your hand: {} ({})\nSilverwolf's hand: {} ({})", 
                    format_hand(&player_hand), calculate_hand(&player_hand),
                    format_hand(&dealer_hand), calculate_hand(&dealer_hand)
                )))
            .components(vec![])
        ).await?;
    } else {
        while calculate_hand(&dealer_hand) < 17 {
            dealer_hand.push(deck.pop().unwrap());
        }

        let p_score = calculate_hand(&player_hand);
        let d_score = calculate_hand(&dealer_hand);

        let user_id = ctx.author().id.to_string();

        if d_score > 21 || p_score > d_score {
            let (_multi, streak, winnings) = ctx.data().db.record_blackjack_win(&user_id, amount).await?;
            ctx.send(poise::CreateReply::default()
                .embed(serenity::CreateEmbed::new()
                    .color(0x00AA00)
                    .title(format!("You win! You won {} mystic credits!", utils::format(winnings)))
                    .description(format!("Your hand: {} ({})\nSilverwolf's hand: {} ({})\nYou are now on a streak of {}", 
                        format_hand(&player_hand), p_score,
                        format_hand(&dealer_hand), d_score,
                        streak
                    )))
                .components(vec![])
            ).await?;
        } else if p_score < d_score {
            ctx.data().db.record_blackjack_loss(&user_id, amount).await?;
            ctx.send(poise::CreateReply::default()
                .embed(serenity::CreateEmbed::new()
                    .color(0xAA0000)
                    .title(format!("Silverwolf wins! You lost {} mystic credits!", utils::format(amount)))
                    .description(format!("Your hand: {} ({})\nSilverwolf's hand: {} ({})", 
                        format_hand(&player_hand), p_score,
                        format_hand(&dealer_hand), d_score
                    )))
                .components(vec![])
            ).await?;
        } else {
            ctx.data().db.record_blackjack_tie(&user_id, amount).await?;
            ctx.send(poise::CreateReply::default()
                .embed(serenity::CreateEmbed::new()
                    .color(0xFFFF00)
                    .title(format!("No one wins! Nothing happened to your {} mystic credits, boring.", utils::format(amount)))
                    .description(format!("Your hand: {} ({})\nSilverwolf's hand: {} ({})", 
                        format_hand(&player_hand), p_score,
                        format_hand(&dealer_hand), d_score
                    )))
                .components(vec![])
            ).await?;
        }
    }

    Ok(())
}
